lib/Default/MySqlDatabase.class.php

```diff
@@ -14,11 +14,51 @@ class MySqlDatabase {
 	private ?array $dbRecord = null;
 
 	public static function getInstance(): MySqlDatabase {
-		return DiContainer::get(self::class);
+		return self::ensureConnected(DiContainer::get(self::class));
 	}
 
 	public static function getNewInstance(): MySqlDatabase {
-		return DiContainer::make(self::class);
+		return self::ensureConnected(DiContainer::make(self::class));
+	}
+
+	/**
+	 * Verifies that the MySqlDatabase instance has a valid $dbConn instance,
+	 * and rewires the DI container to use freshly instantiated instances in the event
+	 * of disconnect from the database.
+	 * @param MySqlDatabase $mysqlDatabase
+	 * @return MySqlDatabase
+	 * @throws \DI\DependencyException
+	 * @throws \DI\NotFoundException
+	 */
+	private static function ensureConnected(MySqlDatabase $mysqlDatabase): MySqlDatabase {
+		if (!isset($mysqlDatabase->dbConn)) {
```

Comment:
Now that I'm thinking about this, I can probably just get away with calling `DiContainer::initialize()` to reset the whole container. If not, then in the future after things are retrieving services through the container, they could end up with a defucnt transient `MySqlDatabase` instance if the connection is ever closed. That would be a lot of manual checks everywhere to make sure reconnect logic is triggered.

I think it's not a big deal since only those CLI apps are using the `close` method, but once the migration starts happening where they don't access db directly, but through domain services, this could be a problem. Maybe just reset the entire DI container for those CLI apps instead of having this reconnect logic.

Reply:
What you have set up now looks like it reproduces the existing behavior. But I agree that avoiding a defunct `MySqlDatabase` instance is probably more robust (especially if future abstraction layers would make misuse easier). However, a few questions I don't have answers for off the top of my head:

What would `DiContainer::initialize()` do to the existing connection? Would it try to close it (if so, might this throw if the connection has timed out)? Or would it leave it open (if so, might this result in a proliferation of open database connections)?

Also, it's important to keep in mind that not all CLI events even need a connection to the database, so I'm not exactly sure where a `DiContainer::initialize()` would go (if we're trying to be as efficient as possible). There are plenty of "reasonable" places to put it, though!

So I'm happy to leave this as-is for now, but if you want to look for a better solution, that's fine with me too.